std/utf.d

```diff
@@ -2420,6 +2469,167 @@ unittest // bugzilla 12923
     }());
 }
 
+/++
+    Returns whether the given range of characters is well-formed Unicode or not.
+
+    Params:
+        range = The range of charaters to validate.
+        index = The index (in code units) where the failed character was (or `0`
+                if the result is `true`).
+  +/
+pragma(inline, true) bool isValidUnicode(R)(R range)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    size_t failedIndex;
+    return _isValidUnicodeImplTrustWrap!false(range, failedIndex);
+}
+
+/// Ditto
+pragma(inline, true) bool isValidUnicode(R)(R range, out size_t failedIndex)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    return _isValidUnicodeImplTrustWrap!true(range, failedIndex);
+}
+
+pragma(inline, true)
+private bool _isValidUnicodeImplTrustWrap(bool getIndex, R)(R range, out size_t failedIndex) @trusted
```

Comment:
This `@trusted` wrapper looks bad to me. There is no indication given as to why the operation it wraps is not already `@safe`, nor of what makes this particular use of it trustworthy.

`_isValidUnicodeImpl()` itself appears to me to be safe, so I'm guessing the problem is the raw pointer access found in some instantiations of `decodeImpl()`. If that is correct, then `decodeImpl()` itself should be reworked to either use safe array operations, or to allow marking those instantiations as `@trusted`.